src/Raven.Client/Documents/AI/AiConversation.cs

```diff
@@ -66,35 +74,91 @@ public void SetUserPrompt(string userPrompt)
     {
-        IMaintenanceOperation<ConversationResult<T>> op;
-        if (string.IsNullOrWhiteSpace(_conversationId))
+        var t = new AiActionContext<TArgs>(_aiOperations, action, errorStrategy);
+        AddAction(actionName, t);
+    }
+
+    private void AddAction(string actionName, IAiActionContext t)
+    {
+        if (_invocations.ContainsKey(actionName))
+            throw new InvalidOperationException($"Action '{actionName}' already exists.");
+
+        _invocations.Add(actionName, t);
+    }
+  
+    public AiAnswer<TAnswer> Run<TAnswer>() => AsyncHelpers.RunSync(() => RunAsync<TAnswer>());
+
+    public async Task<AiAnswer<TAnswer>> RunAsync<TAnswer>(CancellationToken token = default)
+    {
+        while (true)
         {
-            op = new RunConversationOperation<T>(_agentId, _userPrompt, _parameters);
+            var r = await RunAsyncInternal<TAnswer>(token).ConfigureAwait(false);
+            if (r.Status == AiConversationResult.ActionRequired)
+            {
+                if (_actionRequests.Count == 0)
+                    throw new InvalidOperationException("There are no action requests to process.");
+
+                using (var ctx = JsonOperationContext.ShortTermSingleUse())
+                {
+                    for (int i = _actionRequests.Count - 1; i >= 0; i--)
+                    {
+                        var action = _actionRequests[i];
+                        if (_invocations.TryGetValue(action.Name, out var invocation))
+                        {
+                            {
+                                var response = await invocation.ExecuteAsync(ctx, action.Arguments, token).ConfigureAwait(false);
+                                AddActionResponse(action.ToolId, response);
+                            }
+
+                            _actionRequests.RemoveAt(i);
+                        }
+                    }
+                }
+
+                // we responded to all the action requests, so we can continue
+                if (_actionRequests.Count == 0)
+                    continue;
+            }
+
+            return r;
         }
-        else
-        {
-            // we allow to run the conversation only if it is the first run with no user prompt or tool requests
-            // this way we can fetch the pending actions
-            if (_actionRequests != null && string.IsNullOrEmpty(_userPrompt) && _actionResponses.Count == 0)
-                return AiConversationResult.Done;
+    }
 
-            op = new RunConversationOperation<T>(_conversationId, _userPrompt, _actionResponses, _changeVector);
+    private async Task<AiAnswer<TAnswer>> RunAsyncInternal<TAnswer>(CancellationToken token = default)
+    {
+        // we allow to run the conversation only if it is the first run with no user prompt or tool requests
+        // this way we can fetch the pending actions
+        if (_actionRequests != null && string.IsNullOrEmpty(_userPrompt) && _actionResponses.Count == 0)
```

Comment:
```suggestion
        if ( string.IsNullOrEmpty(_userPrompt) && _actionResponses is { Count == 0} )
```

This would also do it, no?
Need to explain that `_actionResponses` is null the very first time